crescent/ext/cooldowns/__init__.py

```diff
@@ -0,0 +1,50 @@
+from typing import Any, Awaitable, Callable, Optional
+
+from hikari import Snowflake
+from pycooldown import FixedCooldown
+
+from crescent import Context, HookResult
+
+__all__ = ("CooldownCallbackT", "BucketCallbackT", "cooldown")
+
+CooldownCallbackT = Callable[[Context, float], Awaitable[Optional[HookResult]]]
+BucketCallbackT = Callable[[Context], Any]
+
+
+def _default_bucket(ctx: Context) -> Snowflake:
+    return ctx.user.id
+
+
+def cooldown(
+    period: float,
+    capacity: int,
+    callback: Optional[CooldownCallbackT] = None,
+    bucket: Optional[BucketCallbackT] = None,
+) -> Callable[[Context], Awaitable[Optional[HookResult]]]:
+    """
+    Ratelimit implementation using a sliding window.
+
+    Args:
+        period:
+            The period of time, in seconds, between cooldown resets.
+        capacity:
+            The amount of times the command can be used within the period.
+        callback:
+            Callback for when a user is ratelimited.
+    """
+    cooldown: FixedCooldown[Any] = FixedCooldown(period, capacity)
+    get_bucket = bucket or _default_bucket
```

Comment:
How about instead of making bucket optional, we just did `bucket: BucketCallbackT = _default_bucket`